bikeshed/update/manifest.py

```diff
@@ -143,28 +145,54 @@ def updateByManifest(path, dryRun=False):
         lastMsgTime = time.time()
         if newPaths:
             say("Updating {0} file{1}...", len(newPaths), "s" if len(newPaths) > 1 else "")
+
+        updateQueue = queue.Queue()
+        success = True
+        # This worker will download from remote, and update success if it fails.
+        def worker():
+            # Needed to update status and progress.
+            nonlocal success
+            nonlocal lastMsgTime
+            while True:
+                (i, filePath) = updateQueue.get()
+                remotePath = ghPrefix + filePath
+                localPath = localizePath(path, filePath)
+                try:
+                    newFile = requests.get(remotePath).text
+                except Exception as e:
+                    warn("Couldn't download file '{0}'.\n{1}", remotePath, e)
+                    success = False
+                try:
+                    dirPath = os.path.dirname(localPath)
+                    if not os.path.exists(dirPath):
+                        os.makedirs(dirPath)
+                    with io.open(localPath, 'w', encoding="utf-8") as fh:
+                        fh.write(newFile)
+                except Exception as e:
+                    warn("Couldn't save file '{0}'.\n{1}", localPath, e)
+                    success = False
+
+                currFileTime = time.time()
+                if (currFileTime - lastMsgTime) >= messageDelta:
+                    say("Updated {0}/{1}...", i+1, len(newPaths))
+                    lastMsgTime = currFileTime
+                updateQueue.task_done()
+
+        # Enqueue all the files to update. i is used for printing progress.
         for i,filePath in enumerate(newPaths):
-            remotePath = ghPrefix + filePath
-            localPath = localizePath(path, filePath)
-            try:
-                newFile = requests.get(remotePath).text
-            except Exception as e:
-                warn("Couldn't download file '{0}'.\n{1}", remotePath, e)
-                return False
-            try:
-                dirPath = os.path.dirname(localPath)
-                if not os.path.exists(dirPath):
-                    os.makedirs(dirPath)
-                with io.open(localPath, 'w', encoding="utf-8") as fh:
-                    fh.write(newFile)
-            except Exception as e:
-                warn("Couldn't save file '{0}'.\n{1}", localPath, e)
-                return False
-
-            currFileTime = time.time()
-            if (currFileTime - lastMsgTime) >= messageDelta:
-                say("Updated {0}/{1}...", i+1, len(newPaths))
-                lastMsgTime = currFileTime
+            updateQueue.put((i, filePath))
+
+        # Create as many threads as we can.
+        for i in range(len(os.sched_getaffinity(0))):
```

Comment:
Whether or not you use threads, the number of parallel HTTP fetches shouldn't be based on the number of usable CPUs. HTTP handling isn't CPU-bound, and Python threads run under a global lock that prevents them from using more than 1 net CPU anyway. (That's why `multiprocessing` exists, but again, because HTTP handling isn't CPU-bound, it won't help here.)

Reply:
> Whether or not you use threads, the number of parallel HTTP fetches shouldn't be based on the number of usable CPUs.

Sure, this is merely a heuristic. If I have 10 cores, I should be able to run up to 10 threads that do blocking work. It's a quick way of running blocking network calls in parallel.

> Python threads run under a global lock that prevents them from using more than 1 net CPU anyway.

I don't know how the GIL interacts here, but from my testing, I see a speedup. If the GIL prevents these threads from being parallelized, why do we see a speedup?

> HTTP handling isn't CPU-bound, it won't help here

Yea, it's a network call. But in this case the network call blocks, you can imagine it's a CPU busy wait (for network response).

Reply:
Sorry, I didn't write that clearly. The GIL prevents multiple Python threads from using CPU concurrently, but it gets released while a Python thread is waiting on an OS function, which is what happens while one of these HTTP fetches waits for the response to come back. That's *not* a busy-wait.

So, even with just 1 core, you can efficiently run lots of threads here to get that many HTTP fetches to be sent in parallel. They can't actually process results in parallel, but that's fine because processing results takes much less wall-time than waiting for the server to respond.